src/Propel/Common/Config/Loader/IniFileLoader.php

```diff
@@ -144,37 +138,40 @@ private function parseSection(array $section): array
      */
     private function parseKey(string $key, $value, array &$config): void
     {
-        if (strpos($key, $this->nestSeparator) !== false) {
-            [$subKey, $restKey] = explode($this->nestSeparator, $key, 2);
-
-            if ($subKey === '' || $restKey === '') {
+        if (is_string($value)) {
+            if (strlen($value) <= 5 && in_array(strtolower($value), ['true', 'false'], true)) {
+                $value = (strtolower($value) === 'true');
+            } elseif ($value === (string)(int)$value) {
+                $value = (int)$value;
+            } elseif ($value === (string)(float)$value) {
+                $value = (float)$value;
+            }
+        }
+        $subKeys = explode($this->nestSeparator, $key);
+        $subConfig = &$config;
+        $lastIndex = count($subKeys) - 1;
+        foreach ($subKeys as $index => $subKey) {
+            if ($subKey === '') {
                 throw new IniParseException(sprintf('Invalid key "%s"', $key));
             }
-            if (!isset($config[$subKey])) {
-                if ($subKey === '0' && !empty($config)) {
-                    $config = [$subKey => $config];
+            if ($index === $lastIndex) {
+                $subConfig[$subKey] = $value;
+
+                break;
+            }
+            if (!isset($subConfig[$subKey])) {
+                if ($subKey === '0' && !empty($subConfig)) {
+                    $subConfig = [$subKey => $subConfig];
```

Comment:
Ah, another mind-bender left to us by those that came before us. I don't understand what this `if` tries to do.

What does it mean, when `$subConfig` is not empty here? Does it mean that there was a parameter definition like
```
domain.section = a
domain.section.0 = 'b'
```
or is it for:
```
domain.section.1.value = a
domain.section.0.value = 'b'
```
But I am not sure if the latter would work. Wouldn't we end up with:
```php
['domain' => ['section' => [
  0 => [
    1 => ['value' => 'a'],
    'value' => 'b'
  ]
]]]
```
And I am not sure if a parameter like `section.0 = 42` leads to `['section' => [42]]`.

The problem is, even if my guesses are wrong, and it actually does work, for me, I cannot see it. It's just too much voodoo.

I don't mind untangling this, but we might be better off doing it in its own PR, as it is probably going to be complicated. My guess is that the whole method needs to be rewritten with tests. What do you think?

Reply:
According to the tests, the "0" is meant to transform this:
```
foo = bar
0.babar = baz
```
into
```php
['0' => ['foo' => 'bar', 'babar' => 'baz']]
```
Is that as weird as I think it is?

Reply:
@mringler Yes this is weird, one should instead write like this.

```ini
0.foo = bar
0.babar = baz
```
or
```ini
[0]
foo = bar
babr = baz
```

> I don't mind untangling this, but we might be better off doing it in its own PR, as it is probably going to be complicated. My guess is that the whole method needs to be rewritten with tests. What do you think?

Agreed.